### ISSUE TITLE
Non-reproducibility across restarts for VERTEX_SHEAR=True

### ISSUE THREAD

**Issue by wfcooke** [NOAA-GFDL/MOM6#850]
With the SPEAR setup (1 degree ocean, symmetric) I am seeing differences in answers across restarts. 1x8 day vs 2x4 day.
For VERTEX_SHEAR=False I reproduce old answers.

**Comment by Hallberg-NOAA** [comment]
Do the results reproduce across restarts when run in non-symmetric memory mode?

**Comment by wfcooke** [comment]
@Hallberg-NOAA : Sorry Bob, I should have looked at the code rather then relying on memory. I am using the dynamic path (not dynamic_symmetric) of the code.

**Comment by wfcooke** [comment]
@Hallberg-NOAA @adcroft 
I ran the symmetric code with VERTEX_SHEAR=True and DEBUG and found that u and v have different checksums immediately after restart.

There is an issue with the restart file.
@MJHarrison-GFDL had me take a look at MOM.res.nc and there is a difference in the redundant points.

yes? list/k=1/precision=6 u[j=35:50,i=361],u[j=35:50,i=1]
             DATA SET: ./MOM.res.nc
             LATITUDE: 68.2S to 61.5S
             Z (meter): 1
             T (days): 0.25
 Column  1: U[X=60E] is Zonal velocity (m s-1)
 Column  2: U[X=60E(-300)] is Zonal velocity (m s-1)
                     U       U
68.06S / 35:  0.0000000  0.0000000
67.68S / 36:  0.0000000  0.0000000
67.3S  / 37: -0.0026159 -0.0026159
66.91S / 38: -0.0475604 -0.0475604
66.51S / 39: -0.0033971 -0.0033971
66.11S / 40:  0.0259861  0.0259861
65.7S  / 41:  0.0299010  0.0299010
65.29S / 42:  0.0253395  0.0253395
64.87S / 43:  0.0261799  0.0261799
64.44S / 44:  0.0154302  0.0154302
64S    / 45:  0.0051715  0.0051713
63.56S / 46: -0.0034555 -0.0034556
63.11S / 47:  0.0030681  0.0030681
62.66S / 48: -0.0033926 -0.0033925
62.2S  / 49: -0.0070073 -0.0070025
61.73S / 50: -0.0098893 -0.0098852

(j=0:34 were all zeroes)

Will

**Comment by wfcooke** [comment]
I should note that with my old (Control case) settings (i.e the above with VERTEX_SHEAR=False, I do not see any difference across restarts, so it looks like VERTEX_SHEAR=True is tickling something.

**Comment by wfcooke** [comment]
![vertex_log_udiff](https://user-images.githubusercontent.com/5122720/45179259-00dedc80-b1e6-11e8-8e98-6076e6ddede4.gif)
![vertex_udiff](https://user-images.githubusercontent.com/5122720/45179260-00dedc80-b1e6-11e8-9732-5046d21989b4.gif)

Attached are plots of the difference and log(abs(difference)) between the i=1 and i=361 point.

**Comment by wfcooke** [comment]
@Hallberg-NOAA @adcroft @MJHarrison-GFDL 
I ran different jobs with 144x2 cores on the ATM side and 63, 64 and 96 cores on the OCN side. 
Checksums at the end of the run indicate different answers. I also tried 288x2 on the ATM with 180 on the OCN and got a different answer. uo and vo are different after 8 days (and less than that)
With VERTEX_SHEAR=False, answers reproduce across these core counts.

**Comment by nikizadehgfdl** [comment]
I tried OM4_025 + VERTEX_SHEAR=True with dev/gfdl/2018.08.25.
It did not reproduce across a restart. It did not reproduce across layout change either.
OM4_025 itself passes these tests fine.

**Comment by Hallberg-NOAA** [comment]
I just introduced PR #1184, which corrects the problems with VERTEX_SHEAR reproducing across layouts in the recently updated ice_ocean_SIS2/SIS2_bergs/cgrid test case, which both has ice and uses VERTEX_SHEAR. If this still an active issue for you, @wfcooke, it would be helpful if you or @nikizadehgfdl could test whether you are still having issues after this PR.

If you are testing the reprroducability across restarts, please note that for these cases to reproduce across restarts, it might be necessary to use symmetric memory and to make sure that the restart files do include the prior shear viscosities in the points along both edges of the domain.  The previous turbulent viscosities are only used as first guesses to speed the convergence of the iterations, so disabling this would also be a way to break the dependencies on the restart files, although it could add a lot to the computational cost of this routine.

**Comment by nikizadehgfdl** [comment]
I tried the OMIP4p25 with  VERTEX_SHEAR again with dev/gfdl. It did not reproduce across a restart (1x2day != 2x1day) in either symmetric or non-symmetric compilation, with or without mask_table. This experiment doe not have an initial condition and starts from WOA. 
Location of restarts for reference
```
/lustre/f2/scratch/Niki.Zadeh/FMS2019.01.03_mom6_20200929_sym/OM4p25_IAF_BLING_csf_rerun_VERTEX_SHEAR/ncrc4.intel18-prod/archive/1x0m2d_864x1o/restart/17080103.tar 
/lustre/f2/scratch/Niki.Zadeh/FMS2019.01.03_mom6_20200929/OM4p25_IAF_BLING_csf_rerun_VERTEX_SHEAR/ncrc4.intel18-prod/archive/2x0m1d_864x1o/restart/17080103.tar

Files left in /lustre/f2/scratch/Niki.Zadeh/tmp/ardiff.gaea13.35916.4NHPr2zL7l

```

**Comment by Hallberg-NOAA** [comment]
I believe that PR#1221 will finally address this issue.  OM_1deg is now reproducing across PEs with VERTEX_SHEAR=True, and both the OM_1deg and OM4p25 with VERTEX shear reproduce across restarts for (2x1day = 2days) and (2x12hr = 24 hr), respectively.  @nikizadehgfdl and @wfcooke, if you have additional tests you would like to run to verify that this issue is corrected after PR#1221 is merged, please do so, and when you are satisfied I will be closing this issue.
